src/netstandard/ref/System.Data.Common.cs

```diff
@@ -466,9 +474,11 @@ public abstract partial class DbDataReader : System.MarshalByRefObject, System.C
         public abstract int RecordsAffected { get; }
         public virtual int VisibleFieldCount { get { throw null; } }
         public virtual void Close() { }
+        public virtual System.Threading.Tasks.Task CloseAsync(System.Threading.CancellationToken cancellationToken = default) { throw null; }
```

Comment:
One note: it seems a little weird to accept a cancellation token here. The reasoning for not including a CancellationToken in DisposeAsync is as follows:

```C#


    DisposeAsync accepting a CancellationToken: while in theory it makes sense that anything async can be canceled, disposal is about cleanup, closing things out, free'ing resources, etc., which is generally not something that should be canceled; cleanup is still important for work that's canceled. The same CancellationToken that caused the actual work to be canceled would typically be the same token passed to DisposeAsync, making DisposeAsync worthless because cancellation of the work would cause DisposeAsync to be a nop. If someone wants to avoid being blocked waiting for disposal, they can avoid waiting on the resulting ValueTask, or wait on it only for some period of time.
```

Reply:
Thanks for the comment! It seems like there may be cause to treat CloseAsync a bit differently from DisposeAsync.

First, it's less likely that the same CancellationToken that caused the work to be canceled would end up being passed to CloseAsync (assuming the standard async using pattern etc.).

Also, at least for DbConnection, the same object can closed and opened several times, so closing isn't really about terminal cleanup (object continues to be usable afterwards), but possibly a bit more similar to a standard I/O operation. That's admittedly less true for DbDataReader which can't be reopened once closed - although some properties on it remain alive (and in some cases even only accessible) after close and before dispose.

@divega @terrajobst what do you think? Will hold off on merging this for a bit.

Reply:
@roji, I came to the exact same conclusions as you without reading your comment :smile: In particular about the difference between CloseAsync and DisposeAsync.

But then I looked at some of the code we have already put together in EF Core to react to to see if I could find any interesting anecdotal information.  It turns out, we have a few different cases that made me change my mind completely. 

TL;DR: I think I found proof that @agocke is right, and **having CancellationToken on CloseAsync is actually a pit of failure**

Here are the examples:

- In [MigrationCommandExecutor.ExecuteNonQueryAsync](https://github.com/aspnet/EntityFrameworkCore/blob/master/src/EFCore.Relational/Migrations/Internal/MigrationCommandExecutor.cs#L142), we are using DbConnection.CloseAsync(CancellationToken) as part of finally block, and we are effectively passing the same CancellationToken we passed before to DbCommand.ExecuteNonQueryAsync.   It seems difficult to figure out that this is wrong.

- In [RelationalTransaction.ClearTransactionAsync](https://github.com/aspnet/EntityFrameworkCore/blob/master/src/EFCore.Relational/Storage/RelationalTransaction.cs#L352) we call CloseAsync(CancellationToken) on the connection, also passing the same cancellation token we pass to the previous call that does some work. 

- In [BatchExecutor.ExecuteAsync](https://github.com/aspnet/EntityFrameworkCore/blob/master/src/EFCore.Relational/Update/Internal/BatchExecutor.cs#L168) same problem.

- In all the execute methods in [RelationalCommand](https://github.com/aspnet/EntityFrameworkCore/blob/master/src/EFCore.Relational/Storage/RelationalCommand.cs) same problem.

- In [SqlServerDatabaseCreator.ExistsAsync](https://github.com/aspnet/EntityFrameworkCore/blob/master/src/EFCore.SqlServer/Storage/Internal/SqlServerDatabaseCreator.cs#L242) we call IRelationalConnection.CloseAsync without passing the cancellation token! I am not sure if we should call it an oversight :smile:

@roji, @ajcvickers I will create an issue in EF Core to track removing the cancellation token. If you agree, can we get this changes into .NET Core and this PR?

Reply:
Related: https://github.com/aspnet/EntityFrameworkCore/issues/16353.

Reply:
On DbConnection, it seems a little problematic to have OpenAsync accept a token but not CloseAsync - but I do agree that the pit of failure here outweighs that. Am fine with removing the token from CloseAsync, will wait to hear @ajcvickers's opinion.

Reply:
I'm on the fence, but I'm fine with removing it.

Reply:
> On DbConnection, it seems a little problematic to have OpenAsync accept a token but not CloseAsync

Do you have in mind any specific case in which this is problematic? 

I haven't come up with any case in which because of hoe OpenAsync() is called, we would need CloseAsync() to accept a CancellationToken to achieve the correct behavior. 

Example 1:

``` c#
try
{
    await connection.OpenAsync(ct);
    await command.ExecuteNonQueryAsync(connection, ct);
}
finally
{
    await connection.CloseAsync();
}
```

- If cancellation is triggered after OpenAsync() executes but before ExecuteNoQueryAsync() executes. The latter should then throw, causing the finally block to be executed. Since the connection is open, CloseAsync should close the connection. 

- If cancellation is triggered before OpenAsync() executes, the latter will not open the connection and throw, causing the finally block to be executed. Since the connection isn't open, [CloseAsync should then be a no-op](https://docs.microsoft.com/dotnet/api/system.data.common.dbconnection.close#remarks). 

Example 2:

``` c#
await connection.OpenAsync(ct);
try
{
    await command.ExecuteNonQueryAsync(connection, ct);
}
finally
{
    await connection.CloseAsync();
}
```

If cancellation is triggered before OpenAsync(), the connection should not be opened and the exception should be handled elsewhere. This is the pattern in  

If OpenAsync() implementation is buggy and does not throw if its cancelled, then it becomes equivalent to one of the other cases.

Reply:
The scenarios I have in mind are where DbConnection.CloseAsync isn't used for cleanup (in a finally block), but as a close operation where the connection is expected to be reopened. It's also worth noting that this is valid for non-pooled connections - with pooled connections (the common case) CloseAsync is expected to return synchronously in all but some rare uncommon cases (since it's putting the connection back in the pool).

But I do agree that scenarios where close cancellation is important are niche, and see more potential damage in keeping the token (i.e. when the same cancelled token is accidentally in passed for cleanup). Since @ajcvickers is OK with removing, I can submit a PR to corefx and see whether @stephentoub or @terrajobst have any objections.

Reply:
Submitted PR to corefx for token removal: https://github.com/dotnet/corefx/pull/39070